### PR TITLE
add ability to pass in all config values as environment variables.

### PR DESCRIPTION
**Pull request by hramezani**
add environment variables that starts with `GUNICORN_` like `GUNICORN_WORKERS` as settings.
skip variables that does not start with `GUNICORN_`, and also skip variable that not in `settings` like `GUNICORN_FOO`

### CODE REVIEW

File: gunicorn/config.py [hunk 1]
```diff
@@ -69,6 +69,19 @@ def set(self, name, value):
             raise AttributeError("No configuration setting for: %s" % name)
         self.settings[name].set(value)
 
+    def parse_env_vars(self):
+        env_vars = {}
+        for k,v in self.env_orig.items():
+            if k.startswith('GUNICORN_'):
+                try:
+                    name = k.split('GUNICORN_', 1)[1].lower()
```
Comment (tilgovi) [inline gunicorn/config.py]: I don't think this can ever fail if `k.startswith('GUNICORN_')`.

File: gunicorn/app/base.py [hunk 2]
```diff
@@ -154,6 +154,11 @@ def load_config(self):
             if default_config is not None:
                 self.load_config_from_file(default_config)
 
+        # Load up environment configuration
+        env_vars = self.cfg.parse_env_vars()
+        for k,v in env_vars.items():
```
Comment (tilgovi) [inline gunicorn/app/base.py]: small linting nit here, space between `,v`.

File: gunicorn/config.py [hunk 3]
```diff
@@ -69,6 +69,19 @@ def set(self, name, value):
             raise AttributeError("No configuration setting for: %s" % name)
         self.settings[name].set(value)
 
+    def parse_env_vars(self):
+        env_vars = {}
+        for k,v in self.env_orig.items():
```
Comment (tilgovi) [inline gunicorn/config.py]: Same nitpick here about spacing.

File: gunicorn/config.py [hunk 4]
```diff
@@ -69,6 +69,19 @@ def set(self, name, value):
...
+            if k.startswith('GUNICORN_'):
+                try:
+                    name = k.split('GUNICORN_', 1)[1].lower()
+                except:
+                    continue
+                if not name or name not in self.settings:
```
Comment (tilgovi) [inline gunicorn/config.py]: `name not in self.settings` should be sufficient, no?

**Comment by hramezani** [comment]
@tilgovi  thanks, i fix problems.

**Comment by berkerpeksag** [comment]
I also like the idea, thanks! I haven't reviewed the code yet, but we should document all `GUNICORN_` env variables before merging this.

File: tests/test_config.py [hunk 5]
```diff
@@ -285,3 +285,25 @@ def test_always_use_configured_logger():
...
+    os.environ["GUNICORN_WORKERS"] = "4"
+    with AltArgs():
+        app = NoConfigApp()
+        assert app.cfg.workers == 4
+
+def test_not_loading_enviroment_variables_config():
+    os.environ["GUNICORN_FOO"] = "BAR"
+    with AltArgs():
+        app = NoConfigApp()
+        try:
```
Comment (berkerpeksag) [inline tests/test_config.py]: This can be written as:

```py
with pytest.raises(AttributeError):
    app.cfg.too
```

File: tests/test_config.py [hunk 6]
```diff
@@ -285,3 +285,25 @@ def test_always_use_configured_logger():
     c.set('statsd_host', 'localhost:12345')
     # still uses custom logger over statsd
     assert c.logger_class == MyLogger
+
+def test_load_enviroment_variables_config():
+    os.environ["GUNICORN_WORKERS"] = "4"
```
Comment (berkerpeksag) [inline tests/test_config.py]: We need to unset these environment variables. I don't know what's the best practice to do this in pytest though (I'd suggest writing a context manager if we were using unittest -- perhaps you need a fixture?)

**Comment by hramezani** [comment]
@berkerpeksag Thanks, 
this PR accept `KNOWN_SETTINGS` in this format `GUNICORN_setting_name`. setting names are in [Settings](http://docs.gunicorn.org/en/stable/settings.html). for example `GUNICORN_bind`, `GUNICORN_worker_class`, ...
Is this a good idea that a write a paragraph in settings page and mention that we can change all of settings in this page by setting environment  variables with abov format?

**Comment by benoitc** [comment]
What is the use case for this change? Can you elaborate about it?

**Comment by hramezani** [comment]
@benoitc i saw [`Ability to pass in all config values as environment variables`](https://github.com/benoitc/gunicorn/issues/1360) in issue list and @tilgovi`  commented:

> I don't see any reason not to support it. Pull requests welcome.

finally i decide to do this.

with this change we can change gunicorn settings with environment variables.

**Comment by tilgovi** [comment]
There are times, especially in these days of containerization, where it is much easier to change an environment variable than to change the command line invocation or a configuration file.

**Comment by tilgovi** [comment]
We can do this in a separate PR, but I'm wondering if there is any reason to clear these keys from the environment when we spawn workers. I'm just wondering if any settings might contain information that we should try to prevent from leaking to workers and possibly then into a response.

**Comment by hramezani** [comment]
@benoitc @tilgovi 
Is there anything that I do?

**Comment by berkerpeksag** [comment]
> Is this a good idea that a write a paragraph in settings page and mention that we can change all of settings in this page by setting environment variables with abov format?

It would be nice to add something like "Environment variable: `GUNICORN_FOO`" (or just plain `GUNICORN_FOO`) to every setting. For example, let's take a look at `bind`: http://docs.gunicorn.org/en/stable/settings.html#bind With "-b ADDRESS, --bind ADDRESS", we know that we can pass `-b` or `--bind` to `gunicorn`. We can document the env variable in the same way:

```
bind

* -b ADDRESS, --bind ADDRESS (CLI)
* GUNICORN_BIND (env variable)
* ['127.0.0.1:8000'] (default value)
```

**Comment by hramezani** [comment]
@berkerpeksag thanks for reply, I add documents.

**Comment by berkerpeksag** [comment]
Thanks, `docs/source/settings.rst` is automatically generated. I prefer modifying https://github.com/benoitc/gunicorn/blob/master/docs/gunicorn_ext.py#L39 instead of manually adding these names. You can then run `make -C docs html` to update `settings.rst` (note that `make html` will make some unrelated changes in `settings.rst` -- please discard them before committing your changes)

File: docs/source/settings.rst [hunk 7]
```diff
@@ -17,6 +17,7 @@ config
 ~~~~~~
 
 * ``-c CONFIG, --config CONFIG``
+* ``GUNICORN_CONFIG (env variable)``
```
Comment (berkerpeksag) [inline docs/source/settings.rst]: "(env variable)" can be removed. I used it as an annotation in my example :)

**Comment by hramezani** [comment]
@berkerpeksag Thanks, whta is your opinion about adding a property like `cli` with name `env_var` to Setting class? this can be used for following purpose:

1. change the environment variable name(now variables name  are the same as the `name` property).
2. add variables to config if this propery has valid value.
3. disable this functionality for some setting.
4. generate document based on this property.

**Comment by benoitc** [comment]
I'm not sure i'm comfortable with this change, it introduces an extra level of complexity in the configuration, having to document the variables to use is a sign. What about instead having a simple `GUNICORN_CMD_ARGS` (or whatever its name) where you pass the command line arguments that could be parsed like usual? Ie:

`GUNICORN_CMD_ARGS ="--bind=127.0.0.1 --workers=3"` ...

So we wouldn't have to document each changes and could reuse the current code to handle it. Thoughts?

**Comment by hramezani** [comment]
@benoitc Thanks, I change PR as your previous comment.

**Comment by decal** [comment]
Looks even better than the Apache 2.4 envvars support..

File: gunicorn/config.py [hunk 8]
```diff
@@ -69,6 +69,12 @@ def set(self, name, value):
             raise AttributeError("No configuration setting for: %s" % name)
         self.settings[name].set(value)
 
+    def get_gunicorn_env_var(self):
+        env_vars = []
```
Comment (berkerpeksag) [inline gunicorn/config.py]: Unless I'm missing something obvious, `env_vars` is not needed here:

```py
if 'GUNICORN_CMD_ARGS' in self.env_orig:
    return self.env_orig['GUNICORN_CMD_ARGS'].split()
return []
```

File: gunicorn/config.py [hunk 9]
```diff
@@ -69,6 +69,12 @@ def set(self, name, value):
...
+        if 'GUNICORN_CMD_ARGS' in self.env_orig:
+            env_vars = self.env_orig['GUNICORN_CMD_ARGS'].split()
```
Comment (berkerpeksag) [inline gunicorn/config.py]: Did you consider using `shlex.split()` here?

File: gunicorn/config.py [hunk 10]
```diff
@@ -69,6 +69,12 @@ def set(self, name, value):
             raise AttributeError("No configuration setting for: %s" % name)
         self.settings[name].set(value)
 
+    def get_gunicorn_env_var(self):
```
Comment (berkerpeksag) [inline gunicorn/config.py]: I'm not a native English speaker, but perhaps we can rename it to something like `get_cmd_args_from_env`? `get_gunicorn_env_var` sounds too generic to me. Ping @tilgovi :)

File: gunicorn/app/base.py [hunk 11]
```diff
@@ -154,6 +154,17 @@ def load_config(self):
             if default_config is not None:
                 self.load_config_from_file(default_config)
 
+        # Load up environment configuration
+        env_vars = self.cfg.get_gunicorn_env_var()
+        if env_vars:
+            env_args = parser.parse_args(env_vars)
+            for k, v in env_args.__dict__.items():
```
Comment (berkerpeksag) [inline gunicorn/app/base.py]: Wouldn't `vars(env_args).items()` be better?

File: tests/test_config.py [hunk 12]
```diff
@@ -285,3 +285,15 @@ def test_always_use_configured_logger():
...
     # still uses custom logger over statsd
     assert c.logger_class == MyLogger
+
+def test_load_enviroment_variables_config():
+    os.environ["GUNICORN_CMD_ARGS"] = "--workers=4"
+    with AltArgs():
+        app = NoConfigApp()
+        assert app.cfg.workers == 4
+
+def test_cli_overrides_enviroment_variables_module():
```
Comment (berkerpeksag) [inline tests/test_config.py]: Tests are look great, thank you! :) Could you also add a test that passes an invalid config option (e.g `os.environ["GUNICORN_CMD_ARGS"] = "--monty-python"`?

**Comment by hramezani** [comment]
@berkerpeksag Thanks, i fix all your comments.

**Comment by berkerpeksag** [comment]
Looks great, thanks! I think you forgot to address two of my previous comments:

* We need to cleanup the env variables we set in tests. We can do this in the following way:

  ```py
  def test_cli_overrides_enviroment_variables_module(monkeypatch):
      monkeypatch.setenv("GUNICORN_CMD_ARGS", "--workers=4")
      # test code
  ```
* `GUNICORN_CMD_ARGS` needs to be documented. We can mention it in `docs/source/run.rst` and `docs/gunicorn_ext.py`. We can leave this to a separate PR if you want.

File: tests/test_config.py [hunk 13]
```diff
@@ -285,3 +285,21 @@ def test_always_use_configured_logger():
...
+    os.environ["GUNICORN_CMD_ARGS"] = "--workers=4"
+    with AltArgs():
+        app = NoConfigApp()
+        assert app.cfg.workers == 4
+
+def test_invalid_enviroment_variables_config():
+    os.environ["GUNICORN_CMD_ARGS"] = "--foo=bar"
+    with AltArgs():
+        with pytest.raises(SystemExit):
+            app = NoConfigApp()
```
Comment (berkerpeksag) [inline tests/test_config.py]: Style nit: No need to set `NoConfigApp()` to `app`. Just call `NoConfigApp()`.

File: tests/test_config.py [hunk 14]
```diff
@@ -285,3 +285,21 @@ def test_always_use_configured_logger():
     c.set('statsd_host', 'localhost:12345')
     # still uses custom logger over statsd
     assert c.logger_class == MyLogger
+
```
Comment (berkerpeksag) [inline tests/test_config.py]: Style nit: Please add two empty lines after each tests.

**Comment by hramezani** [comment]
@berkerpeksag thanks!, comments fixed.
can you help me for documentation? is this ok to add a note or paragraph in `Settings` and ``

**Comment by berkerpeksag** [comment]
Thanks for the PR, @hramezani! :)

> is this ok to add a note or paragraph in Settings and Running Gunicorn page, and describe that we can use all command line config in env vars in bellow format?
> GUNICORN_CMD_ARGS ="--bind=127.0.0.1 --workers=3"

Yes, that sounds good to me. Let's open another pull request for documentation.

**Comment by hramezani** [comment]
@berkerpeksag  thanks very much for your help. i open another pull request for documentation soon.

**Pull request merged**